plugins/ipfs/util.go

```diff
@@ -0,0 +1,310 @@
+	case metricBwIn:
+		bw, err := GetBW(l)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprint(bw.TotalIn), nil
+	case metricBwOut:
+		bw, err := GetBW(l)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprint(bw.TotalOut), nil
+	default:
+		return "", errors.New("unrecognized metric: " + metric)
+	}
+}
+
+func GetPeerID(l testbedi.Config) (*cid.Cid, error) {
+	icfg, err := l.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	lcfg, ok := icfg.(*config.Config)
+	if !ok {
+		return nil, fmt.Errorf("Error: GetConfig() is not an ipfs config")
+	}
+
+	pcid, err := cid.Decode(lcfg.Identity.PeerID)
+	if err != nil {
+		return nil, err
+	}
+
+	return pcid, nil
+}
+
+func GetMetricList() []string {
+	return []string{metricBwIn, metricBwOut}
+}
+
+func GetMetricDesc(metric string) (string, error) {
+	switch metric {
+	case metricBwIn:
+		return "node input bandwidth", nil
+	case metricBwOut:
+		return "node output bandwidth", nil
+	default:
+		return "", errors.New("unrecognized metric")
+	}
+}
+
+func GetAttrList() []string {
+	return []string{attrID, attrPath}
+}
+
+func GetAttrDesc(attr string) (string, error) {
+	switch attr {
+	case attrID:
+		return "node ID", nil
+	case attrPath:
+		return "node IPFS_PATH", nil
+	default:
+		return "", errors.New("unrecognized attribute")
+	}
+}
+
+func ReadLogs(l testbedi.Libp2p) (io.ReadCloser, error) {
+	addrStr, err := l.APIAddr()
+	if err != nil {
+		return nil, err
+	}
+
+	addr, err := multiaddr.NewMultiaddr(addrStr)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(tperson) ipv6
+	ip, err := addr.ValueForProtocol(multiaddr.P_IP4)
+	if err != nil {
+		return nil, err
+	}
+	pt, err := addr.ValueForProtocol(multiaddr.P_TCP)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s:%s/api/v0/log/tail", ip, pt))
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+type BW struct {
+	TotalIn  int
+	TotalOut int
+}
+
+func GetBW(l testbedi.Libp2p) (*BW, error) {
+	addrStr, err := l.APIAddr()
+	if err != nil {
+		return nil, err
+	}
+
+	addr, err := multiaddr.NewMultiaddr(addrStr)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(tperson) ipv6
+	ip, err := addr.ValueForProtocol(multiaddr.P_IP4)
+	if err != nil {
+		return nil, err
+	}
+	pt, err := addr.ValueForProtocol(multiaddr.P_TCP)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s:%s/api/v0/stats/bw", ip, pt))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var bw BW
+	err = json.NewDecoder(resp.Body).Decode(&bw)
+	if err != nil {
+		return nil, err
+	}
+
+	return &bw, nil
+}
+
+func GetAPIAddrFromRepo(dir string) (string, error) {
+	out, err := ioutil.ReadFile(filepath.Join(dir, "api"))
+	return string(out), err
+}
+
+func SwarmAddrs(l testbedi.Core) ([]string, error) {
+	pcid, err := l.PeerID()
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := l.RunCmd(context.TODO(), nil, "ipfs", "swarm", "addrs", "local")
```

Comment:
Again, taking a context in `SwarmAddrs` would allow us to feed it into this function here.